src/schema.rs

```diff
@@ -0,0 +1,569 @@
+    /// For `object` types, only the fields `"type"` and `"properties"` are allowed.
+    /// If any other fields are present, an error is returned.
+    ///
+    /// This strictness ensures that unsupported or misspelled fields in schemas are caught early,
+    /// preventing silent misinterpretation of the schema. The function also validates that
+    /// required fields are present and of the correct type, and that all object property keys are strings.
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// - Required fields are missing or of the wrong type
+    /// - Extra fields are present in `array`, `set`, or `object` schemas
+    /// - Object property keys are not strings
+    /// - An unknown type is encountered
+    fn from_value_impl(schema: Value) -> Result<Self, alloc::string::String> {
+        match schema["type"].as_string() {
+            Ok(t) => match t.as_ref() {
+                "array" | "set" => {
+                    // Enforce strict schema: only allow "type" and "items" for array/set,
+                    // and only "type" and "properties" for object.
+                    // Any extra fields will result in an error to catch schema mistakes early.
+                    if let Value::Object(ref map) = schema {
+                        for key in map.keys() {
+                            if let Value::String(ref s) = key {
+                                if s.as_ref() != "type" && s.as_ref() != "items" {
+                                    return Err(format!("'{t}' type: unexpected field '{s}'"));
+                                }
+                            }
+                        }
+                    }
+
+                    let items: Type = match &schema["items"] {
+                        obj @ Value::Object { .. } => Type::from_value_impl(obj.clone())?,
+                        Value::Undefined => {
+                            return Err("'items' is required".into());
+                        }
+                        _ => {
+                            return Err("'items' should be an object".into());
+                        }
+                    };
+                    if t.as_ref() == "array" {
+                        Ok(Type::Array {
+                            items: Rc::new(items),
+                        })
+                    } else {
+                        Ok(Type::Set {
+                            items: Rc::new(items),
+                        })
+                    }
+                }
+                "boolean" => Ok(Type::Boolean),
+                "integer" => Ok(Type::Integer),
+                "null" => Ok(Type::Null),
+                "number" => Ok(Type::Number),
+                "object" => match &schema["properties"] {
+                    Value::Object(props) => {
+                        // Enforce strict schema: only allow "type" and "properties" for object.
+                        // Any extra fields will result in an error to catch schema mistakes
```

Comment:
Is this meant to parse ARM resource payloads? If so, there are other top level fields we need to account for as well